File: lib/core/card/card_type.dart
import 'package:landlords_3/domain/entities/poker_model.dart';
import 'package:landlords_3/core/card/card_utils.dart';

enum CardTypeEnum {
  single, // 单张
  pair, // 对子
  three, // 三张
  threeWithOne, // 三带一
  threeWithTwo, // 三带二
  straight, // 顺子
  straightPair, // 连对
  plane, // 飞机
  planeWithWings, // 飞机带翅膀
  bomb, // 炸弹
  rocket, // 火箭
  fourWithTwo, // 四带二
  fourWithTwoPair, // 四带两对
  invalid, // 不合法的牌型
}

class CardType {
  static CardTypeEnum getType(List<PokerModel> cards) {
    if (cards.isEmpty) return CardTypeEnum.invalid;

    final sortedCards = CardUtils.sortCards(cards);
    final length = sortedCards.length;

    // 火箭
    if (length == 2 &&
        sortedCards[0].value == CardValue.jokerBig &&
        sortedCards[1].value == CardValue.jokerSmall) {
      return CardTypeEnum.rocket;
    }

    // 统计每张牌的数量
    final counts = <CardValue, int>{};
    for (final card in sortedCards) {
      counts[card.value] = (counts[card.value] ?? 0) + 1;
    }

    // 炸弹
    if (counts.length == 1 && counts.values.first == 4) {
      return CardTypeEnum.bomb;
    }

    // 单张
    if (length == 1) {
      return CardTypeEnum.single;
    }

    // 对子
    if (length == 2 && counts.length == 1 && counts.values.first == 2) {
      return CardTypeEnum.pair;
    }

    // 三张
    if (length == 3 && counts.length == 1 && counts.values.first == 3) {
      return CardTypeEnum.three;
    }

    // 顺子
    if (_isStraight(sortedCards)) {
      return CardTypeEnum.straight;
    }

    // 连对
    if (_isStraightPair(sortedCards)) {
      return CardTypeEnum.straightPair;
    }

    // 飞机
    if (_isPlane(sortedCards)) {
      return CardTypeEnum.plane;
    }

    // 三带一
    if (length == 4 &&
        counts.length == 2 &&
        counts.containsValue(3) &&
        counts.containsValue(1)) {
      return CardTypeEnum.threeWithOne;
    }

    // 三带二
    if (length == 5 &&
        counts.length == 2 &&
        counts.containsValue(3) &&
        counts.containsValue(2)) {
      return CardTypeEnum.threeWithTwo;
    }

    // 飞机带翅膀
    if (_isPlaneWithWings(sortedCards)) {
      return CardTypeEnum.planeWithWings;
    }

    // 四带二
    if (length == 6 &&
        counts.length == 3 &&
        counts.containsValue(4) &&
        counts.containsValue(1)) {
      return CardTypeEnum.fourWithTwo;
    }

    // 四带两对
    if (length == 8 &&
        counts.length == 3 &&
        counts.containsValue(4) &&
        counts.containsValue(2)) {
      return CardTypeEnum.fourWithTwoPair;
    }

    return CardTypeEnum.invalid;
  }

  // 判断是否是顺子
  static bool _isStraight(List<PokerModel> cards) {
    if (cards.length < 5 || cards.length > 12) return false; // 顺子长度必须在 5-12 之间
    for (int i = 0; i < cards.length - 1; i++) {
      if (cards[i].value == CardValue.two ||
          cards[i].value == CardValue.jokerBig ||
          cards[i].value == CardValue.jokerSmall)
        return false; // 顺子不能包含 2 和大小王
      if (CardUtils.getCardWeight(cards[i]) -
              CardUtils.getCardWeight(cards[i + 1]) !=
          1)
        return false; // 必须是连续的
    }
    return true;
  }

  // 判断是否是连对
  static bool _isStraightPair(List<PokerModel> cards) {
    if (cards.length < 6 || cards.length % 2 != 0)
      return false; // 连对长度必须大于等于 6 且是偶数
    for (int i = 0; i < cards.length - 2; i += 2) {
      if (cards[i].value != cards[i + 1].value) return false; // 必须是对子
      if (cards[i].value == CardValue.two ||
          cards[i].value == CardValue.jokerBig ||
          cards[i].value == CardValue.jokerSmall)
        return false; // 连对不能包含 2 和大小王
      if (CardUtils.getCardWeight(cards[i]) -
              CardUtils.getCardWeight(cards[i + 2]) !=
          1)
        return false; // 必须是连续的
    }
    return true;
  }

  // 判断是否是飞机
  static bool _isPlane(List<PokerModel> cards) {
    if (cards.length < 6 || cards.length % 3 != 0)
      return false; // 飞机长度必须大于等于 6 且是 3 的倍数
    final counts = <CardValue, int>{};
    for (final card in cards) {
      counts[card.value] = (counts[card.value] ?? 0) + 1;
    }
    if (counts.values.any((count) => count != 3)) return false; // 必须都是三张
    final cardValues = counts.keys.toList();
    cardValues.sort(
      (a, b) => CardUtils.getCardWeightByValue(
        b,
      ).compareTo(CardUtils.getCardWeightByValue(a)),
    );
    for (int i = 0; i < cardValues.length - 1; i++) {
      if (CardUtils.getCardWeightByValue(cardValues[i]) -
              CardUtils.getCardWeightByValue(cardValues[i + 1]) !=
          1)
        return false; // 必须是连续的
    }
    return true;
  }

  // 判断是否是飞机带翅膀
  static bool _isPlaneWithWings(List<PokerModel> cards) {
    if (cards.length < 8) return false; // 飞机带翅膀长度必须大于等于 8
    final counts = <CardValue, int>{};
    for (final card in cards) {
      counts[card.value] = (counts[card.value] ?? 0) + 1;
    }
    final threeCount = counts.values.where((count) => count == 3).length;
    final oneCount = counts.values.where((count) => count == 1).length;
    final twoCount = counts.values.where((count) => count == 2).length;

    // 飞机带单张
    if (threeCount >= 2 && (oneCount == threeCount || twoCount == threeCount)) {
      // 提取飞机部分
      List<PokerModel> planeCards = [];
      counts.forEach((key, value) {
        if (value == 3) {
          planeCards.addAll(cards.where((card) => card.value == key));
        }
      });
      if (_isPlane(planeCards)) {
        return true;
      }
    }
    return false;
  }
}

File: server_3/src/card/CardType.ts
// src/models/CardType.ts
import { Poker } from '../models/Poker';
import { CardUtils } from './CardUtils';

export enum Suit { hearts, diamonds, clubs, spades, joker }

export enum CardValue {
    three,
    four,
    five,
    six,
    seven,
    eight,
    nine,
    ten,
    jack,
    queen,
    king,
    ace,
    two,
    jokerSmall, // 小王
    jokerBig, // 大王
}

export enum CardTypeEnum {
    Single = 'single',         // 单张
    Pair = 'pair',             // 对子
    Three = 'three',           // 三张
    ThreeWithOne = 'threeWithOne',     // 三带一
    ThreeWithTwo = 'threeWithTwo',     // 三带二
    Straight = 'straight',     // 顺子
    StraightPair = 'straightPair', // 连对
    Plane = 'plane',           // 飞机
    PlaneWithWings = 'planeWithWings', // 飞机带翅膀
    Bomb = 'bomb',             // 炸弹
    Rocket = 'rocket',          // 火箭
    FourWithTwo = 'fourWithTwo',   // 四带二
    FourWithTwoPair = 'fourWithTwoPair', // 四带两对
    Invalid = 'invalid'        // 无效牌型
}

export class CardType {
    static getType(cards: Poker[]): CardTypeEnum {
        if (cards.length === 0) return CardTypeEnum.Invalid;

        const sorted = CardUtils.sortCards(cards);
        const length = sorted.length;

        // 火箭判断
        if (length === 2 &&
            sorted[0].value === CardValue.jokerBig &&
            sorted[1].value === CardValue.jokerSmall) {
            return CardTypeEnum.Rocket;
        }

        // 统计牌值出现次数
        const counts = new Map<CardValue, number>();
        sorted.forEach(card => {
            counts.set(card.value, (counts.get(card.value) || 0) + 1);
        });

        // 炸弹判断
        if (counts.size === 1 && Array.from(counts.values())[0] === 4) {
            return CardTypeEnum.Bomb;
        }

        // 单张
        if (length === 1) return CardTypeEnum.Single;

        // 对子
        if (length === 2 && counts.size === 1 && Array.from(counts.values())[0] === 2) {
            return CardTypeEnum.Pair;
        }

        // 三张
        if (length === 3 && counts.size === 1 && Array.from(counts.values())[0] === 3) {
            return CardTypeEnum.Three;
        }

        // 顺子判断
        if (this.isStraight(sorted)) {
            return CardTypeEnum.Straight;
        }

        // 连对判断
        if (this.isStraightPair(sorted)) {
            return CardTypeEnum.StraightPair;
        }

        // 飞机判断
        if (this.isPlane(sorted)) {
            return CardTypeEnum.Plane;
        }

        // 三带一
        if (length === 4 && counts.size === 2 &&
            Array.from(counts.values()).some(v => v === 3) &&
            Array.from(counts.values()).some(v => v === 1)) {
            return CardTypeEnum.ThreeWithOne;
        }

        // 三带二
        if (length === 5 && counts.size === 2 &&
            Array.from(counts.values()).some(v => v === 3) &&
            Array.from(counts.values()).some(v => v === 2)) {
            return CardTypeEnum.ThreeWithTwo;
        }

        // 飞机带翅膀
        if (this.isPlaneWithWings(sorted)) {
            return CardTypeEnum.PlaneWithWings;
        }

        // 四带二
        if (length === 6 && counts.size === 3 &&
            Array.from(counts.values()).some(v => v === 4) &&
            Array.from(counts.values()).filter(v => v === 1).length === 2) {
            return CardTypeEnum.FourWithTwo;
        }

        // 四带两对
        if (length === 8 && counts.size === 3 &&
            Array.from(counts.values()).some(v => v === 4) &&
            Array.from(counts.values()).filter(v => v === 2).length === 2) {
            return CardTypeEnum.FourWithTwoPair;
        }

        return CardTypeEnum.Invalid;
    }

    private static isStraight(cards: Poker[]): boolean {
        if (cards.length < 5 || cards.length > 12) return false;

        for (let i = 0; i < cards.length - 1; i++) {
            if ([CardValue.two, CardValue.jokerBig, CardValue.jokerSmall]
                .includes(cards[i].value)) return false;

            if (CardUtils.getCardWeight(cards[i].value) -
                CardUtils.getCardWeight(cards[i + 1].value) !== 1) {
                return false;
            }
        }
        return true;
    }

    private static isStraightPair(cards: Poker[]): boolean {
        if (cards.length < 6 || cards.length % 2 !== 0) return false;

        for (let i = 0; i < cards.length; i += 2) {
            if (i + 1 >= cards.length || cards[i].value !== cards[i + 1].value) {
                return false;
            }

            if ([CardValue.two, CardValue.jokerBig, CardValue.jokerSmall]
                .includes(cards[i].value)) return false;

            if (i < cards.length - 2 &&
                CardUtils.getCardWeight(cards[i].value) -
                CardUtils.getCardWeight(cards[i + 2].value) !== 1) {
                return false;
            }
        }
        return true;
    }

    private static isPlane(cards: Poker[]): boolean {
        if (cards.length < 6 || cards.length % 3 !== 0) return false;

        const counts = new Map<CardValue, number>();
        cards.forEach(card => {
            counts.set(card.value, (counts.get(card.value) || 0) + 1);
        });

        if (Array.from(counts.values()).some(v => v !== 3)) return false;

        const sortedValues = Array.from(counts.keys()).sort((a, b) =>
            CardUtils.getCardWeight(b) - CardUtils.getCardWeight(a));

        for (let i = 0; i < sortedValues.length - 1; i++) {
            if (CardUtils.getCardWeight(sortedValues[i]) -
                CardUtils.getCardWeight(sortedValues[i + 1]) !== 1) {
                return false;
            }
        }
        return true;
    }

    private static isPlaneWithWings(cards: Poker[]): boolean {
        if (cards.length < 8) return false;

        const counts = new Map<CardValue, number>();
        cards.forEach(card => {
            counts.set(card.value, (counts.get(card.value) || 0) + 1);
        });

        const threeCounts = Array.from(counts.values())
            .filter(v => v === 3).length;
        const wingCounts = Array.from(counts.values())
            .filter(v => v === 1 || v === 2).length;

        if (threeCounts < 2 || threeCounts !== wingCounts) return false;

        const planeCards = cards.filter(c =>
            counts.get(c.value)! === 3);
        return this.isPlane(planeCards);
    }
}